Link command

fboucquez opened the pull request:
Hi guys, I've created the `link` command.

This command automatically links the nodes keys (vrf and voting keys) to a running network.

The steps are the following:

For non-voting nodes (only VRF keys are linked)

```
symbol-bootstrap start -p testnet -a dual -r
... Assign tokens using faucet to the node signing key in addresses.yml...
... wait until start when using --detached or in another command line...
symbol-bootstrap link
```

For voting nodes (Voting and VRF keys are linked)
```
symbol-bootstrap start -p testnet -a peer -c customPreset.yml -r
... Assign tokens using faucet to the node signing key in  addresses.yml...
... wait until start when using --detached or in another command line...
symbol-bootstrap link
```

customPreset.yml
```
nodes:
    - voting: true
```

The link command does a bunch of validation. Try it without funds!!!

Link command also allows --url to tune the URL to another network node and --maxFee to tune the cost of the operations. You don't even have to have your node running to link the created configuration.


```
symbol-bootstrap config -p testnet -a dual -c customPreset.yml -r
... Assign tokens using faucet to the node signing key in  addresses.yml...
symbol-bootstrap link -u http://api-01.eu-west-1.0.10.0.x.symboldev.network:3000/
```


Some other improvements:
- roles: is automatically created from the flags (we just need the voting flag to enable voting)
- Improved Readme and custom preset examples
- Added custom host configuration in docker-compose.

Note: I'm not creating a voting assembly as there could be 3 more to maintain (API + voting, peer + voting and dual + voting).
I think most people would need to create a custom preset anyway when running testnet nodes.

 -r is to clean up the target folder if you are trying multiple configurations

segfaultxavi reviewed (approved):
I approve the documentation changes, but this PR contains more than just "Readme improvements"... I have no idea what are you doing there...

fboucquez commented:
Hi @segfaultxavi this PR diverged towards the link command. It's pretty cool. Please have a look!!!

segfaultxavi reviewed (changes requested):
I'm going to test the new command now, but meanwhile, here you have some doc comments :)

fboucquez commented on src/service/LinkService.ts:
```diff
@@ -0,0 +1,152 @@
+                    VotingKeyLinkTransaction.create(
+                        Deadline.create(),
+                        votingPublicKey,
+                        1,
+                        26280,
```
Not at an expert on these values.

I think we could allow users to provide custom values via presets. Then bootstrap will use those values to generate the tree files via catapult.tools.votingkey and to generate the right values in the VotingKeyLinkTransaction. This could be a future PR.

fboucquez replied:
Added preset params defaulted to the default ones:)

fboucquez commented on src/service/BootstrapUtils.ts:
```diff
@@ -158,6 +158,10 @@ export class BootstrapUtils {
     }
 
+    public static createVotingKey(votingPublicKey: string): string {
+        return votingPublicKey + '00000000000000000000000000000000';
```
What about padding 0 until 48 bytes are completed? That could mean padding nothing if the voting public key is already 48. Although for that change we would need to improve the voting key generation.

fboucquez commented on cmds/link-testnet.sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+symbol-bootstrap link -t target/testnet
+http://api-01.eu-west-1.0.10.0.x.symboldev.network:3000/
```
Ups. 8pm coding

rg911 reviewed (commented):
should it have a `-u` for unlink?

Also, should it split vrf and voting link with separate flags? e.g. only link vrf (as voting key linked already)

fboucquez commented:
Thanks for the feedback @segfaultxavi @rg911 @gimer 

I've pushed several improvements:

-     Added votingKeyDilution, votingKeyStartEpoch votingKeyEndEpoch preset params linked to voting key generation and tree file.
- Added --unlink command too link
- Added automatic padding to voting keys (removed the hardcoded 000000000000000).
- Voting, signing and VRF keys, transactions, and tree files are generated and announced when required depending on the node role.
- Improved doc around node link and registration
- Added working dir to docker run execution. Removed `bash -c` to docker runs in the config command.

The `link` command is useful not only for voting nodes, also for regular peer nodes. Users would need to call it to register the VRF keys. Xavi will update the guides, `link` does both key registrations.

fboucquez commented on README.md:
`````diff
@@ -129,6 +153,27 @@ transactions:
+  - voting: true
+````
+
+In order to activate the voting node, you will need to announce a `VotingKeyLinkTransaction` signed using the node's signing private, and the voting public key (suffixed with `00000000000000000000000000000000`).
`````
000s removed

fboucquez commented on README.md:
`````diff
@@ -129,6 +153,27 @@ transactions:
         'MyTransaction': 01000000000000...
 ````
 
+##### Enable voting mode in a node:
`````
Added link command to the preset help
